adagrams/game.py

```diff
@@ -1,11 +1,159 @@
+        if count < LETTER_POOL[random_letter]:
+            hand.append(random_letter)
+        
+    return hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    word = word.upper() # Avoid errors due to case sensitivity
+    temp_letter_bank = letter_bank.copy() # Copy of letter_bank to avoid changing the original list
+
+    # Check availability of each letter in word
+    for i in range(len(word)):
+        user_letter = word[i]
+        available = False
+
+        for letter in temp_letter_bank:
+            if user_letter == letter:
+                available = True
+                temp_letter_bank.remove(letter) # Remove the letter from temp_letter_bank if it's already used in word
+        
+        # Return False if a letter in word not in temp_letter_bank
+        if available == False:
+            return False
+
+    return True
+
+LETTER_POINT = {
+    'A': 1, 
+    'B': 3, 
+    'C': 3, 
+    'D': 2, 
+    'E': 1, 
+    'F': 4, 
+    'G': 2, 
+    'H': 4, 
+    'I': 1, 
+    'J': 8, 
+    'K': 5, 
+    'L': 1, 
+    'M': 3, 
+    'N': 1, 
+    'O': 1, 
+    'P': 3, 
+    'Q': 10, 
+    'R': 1, 
+    'S': 1, 
+    'T': 1, 
+    'U': 1, 
+    'V': 4, 
+    'W': 4, 
+    'X': 8, 
+    'Y': 4, 
+    'Z': 10
+    }
 
 def score_word(word):
-    pass
+    word = word.upper()
+
+    # Calculate total points
+    if word:
+        n = len(word)
+        
+        # Additional 8 points if word has 7 or more letters 
+        if n > 6:
+            points = 8
+        else:
+            points = 0
+
+        # Loop through each letter and calculate total points
+        for i in range(n):
+            points += LETTER_POINT[word[i]]
+    else:
+        points = 0
+    
+    return points
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    word_score_list = [] # List of tuples of words and scores
+    score_list = [] # List of scores only
+
+    # Calculate each word's score and add to word_score_list of tuples and score_list
+    for word in word_list:
+        score = score_word(word)
+        score_list.append(score) 
+        word_score_list.append((word,score)) 
+    
+    # Find the maximum score value
+    max_score = max(score_list)
+
+    # Find tuples of words with the highest score and add it to the highest score word list
+    highest_score_word = []
+
+    for word_score in word_score_list:
+        if word_score[1] == max_score:
+            highest_score_word.append(word_score)
+    
+    # Find THE highest score word from the list of highest score words
+    # If there are more than 1, apply the rules to choose the highest score word
+    # Return the first word that has 10 letters, if any
+    for word_score in highest_score_word:
+        if len(word_score[0]) == 10:
+            return word_score
+        
+    # Filter out the list to contain only word(s) with the least letters
+    #highest_score_word_copy = list(highest_score_word)
+    while len(highest_score_word) > 1:
```

Comment:
This while loop ends up being O(n^2) time. Can we think of a better way to do this?

Reply:
Can you explain why is it O(n^2) please?

Reply:
The `.remove()` function takes O(n) time and you are running it for each element in the list, so it's O(n^2).